packages/back-end/src/middleware/authenticateApiRequestMiddleware.ts

```diff
@@ -116,6 +123,19 @@ export default function authenticateApiRequestMiddleware(
 
       const teams = await getTeamsForOrganization(org.id);
 
+      req.context = {
+        org,
+        userId: req.user?.id || "",
+        email: req.user?.email || "",
```

Comment:
When dealing with API requests, it seems like we can't always guarantee there is an authenticated user behind the request. Is that correct? Do we have an idea what the exact cases are? (Is `req.user` actually ever set?) 

It might make sense to create a different Context type to reflect this, such as `ApiReqContext` where fields like userId, email, and username are optional. This will make it more obvious that req.user can be undefined when dealing with API requests. What do you think?

Reply:
@bttf You're correct - there is a case where `req.user` isn't set.

This middleware is for our REST API, and the requested can either use a personal access token (which is tied to a user) or an API Key that isn't tied directly to a user, but instead is just tied to an organization, and has a role (currently either `admin`, or `readonly`. (Note: I do have a PR to add support for a `SCIM` type role, but that hasn't landed yet.)

After writing the above, I wonder if we should just make those fields optional on `ReqContext` since there are valid cases where those fields wouldn't be defined. Then, if we ever need to use those fields, we can run quick checks to ensure their set.

Reply:
>This middleware is for our REST API, and the requested can either use a personal access token (which is tied to a user) or an API Key that isn't tied directly to a user, but instead is just tied to an organization, and has a role (currently either admin, or readonly. (Note: I do have a PR to add support for a SCIM type role, but that hasn't landed yet.)

Thank you for clarifying, this is good to know.

>After writing the above, I wonder if we should just make those fields optional on ReqContext since there are valid cases where those fields wouldn't be defined. Then, if we ever need to use those fields, we can run quick checks to ensure their set.

Can you identify the valid cases where userId, email would not be defined (in a non-REST api situation)? My understanding is that since we're generating the context after the middleware layer using `getContextFromReq`, user and org should always be defined (otherwise `getContextFromReq` would throw an error).

From my experience it would be beneficial to have two contexts defined - one for in-app requests and the other for REST API requests. It would help address above situation and it lets us more accurately define a context object for either situation, which tend to have different requirements.

Reply:
Hmm... looking at the non-REST Api situation, the `ReqContext` object would never have `userId` or `email` not defined. Though, `userName` could be undefined as we don't require a name field for a user. 

I can definitely create the two different types (1 for our internal API where all fields but `userName` are required, and 1 for the rest API where those fields (userName, userId, email) are optional.

My only hesitation (and happy to talk through this in a huddle) is it feels like it won't make a material change in how we consume these two objects. E.G. In ProjectModel, when we're defining the `context` arg for a function, wouldn't we have to declare that it could be either `ReqContext | ApiReqContext`? And in that case `ApiReqContext` would essentially override `ReqContext` so we'd still have to run checks to see if `userId` or `email` was defined before using them?

Perhaps I'm looking at this too narrowly, and the value will come if we use the `ReqContext` object derived from `getContextFromReq` in other places, and that's the point you're making?

Reply:
> ProjectModel, when we're defining the context arg for a function, wouldn't we have to declare that it could be either ReqContext | ApiReqContext? And in that case ApiReqContext would essentially override ReqContext so we'd still have to run checks to see if userId or email was defined before using them?

This is a great point and yes, that is correct. Still, I think we would benefit from being able to use `ReqContext` in places like the router and controller for in-app requests, where we know it's a `ReqContext` (and not an `ApiReqContext`). 

Furthermore, having a method whose context parameter is typed as a union of both contexts (`ReqContext | `ApiRequestContext`)  gives the developer some signal that 'this method is used for both in-app and API requests, so be wary.'  In the future, we can refactor these methods when complexity grows into separate methods for either context.

Overall I think it will help to have two contexts to have that added benefit when we can. What do you think?